src/XRP/network-clients/web-socket-network-client.ts

```diff
@@ -0,0 +1,155 @@
+ * @see https://xrpl.org/rippled-api.html
+ */
+export default class WebSocketNetworkClient {
+  private readonly socket: WebSocket
+  private accountCallbacks: Map<
+    string,
+    (data: WebSocketTransactionResponse) => void
+  >
+  private messageCallbacks: Map<string, (data: WebSocketResponse) => void>
+  private waiting: Map<string, WebSocketResponse | undefined>
+
+  /**
+   * Create a new WebSocketNetworkClient.
+   *
+   * @param webSocketUrl The URL of the rippled node to query.
+   * @see https://xrpl.org/monitor-incoming-payments-with-websocket.html
+   */
+  public constructor(
+    webSocketUrl: string,
+    private handleErrorMessage: (message: string) => void,
+  ) {
+    this.socket = new WebSocket(webSocketUrl)
+    this.accountCallbacks = new Map()
+    this.waiting = new Map()
+
+    this.messageCallbacks = new Map()
+    this.messageCallbacks.set('response', (data: WebSocketResponse) => {
+      const dataStatusResponse = data as WebSocketStatusResponse
+      this.waiting.set(dataStatusResponse.id, data)
+    })
+    this.messageCallbacks.set('transaction', this.handleTransaction)
+
+    this.socket.addEventListener('message', (event) => {
+      const parsedData = JSON.parse(event.data) as WebSocketResponse
+      const dataType = parsedData.type
+      const callback = this.messageCallbacks.get(dataType)
+      if (callback) {
+        callback(parsedData)
+      } else {
+        this.handleErrorMessage(
+          'Received unhandlable message: ' + (event.data as string),
+        )
+      }
+    })
+
+    this.socket.addEventListener('close', (event) => {
+      this.handleErrorMessage(
+        'Web socket disconnected, ' + (event.reason as string),
+      )
+    })
+
+    this.socket.addEventListener('error', (event) => {
+      this.handleErrorMessage('Error: ' + (event as string))
+    })
+  }
+
+  /**
+   * Properly handles incoming transactions from the web socket.
+   *
+   * @param data The web socket response received from the web socket.
+   */
+  private handleTransaction = (data: WebSocketResponse) => {
+    const dataTransaction = data as WebSocketTransactionResponse
+    const destinationAccount = dataTransaction.transaction.Destination
+    const callback = this.accountCallbacks.get(destinationAccount)
+    if (callback) {
+      callback(dataTransaction)
+    } else {
+      throw new XrpError(
+        XrpErrorType.Unknown,
+        'Received a transaction for an account that has not been subscribed to: ' +
+          destinationAccount,
+      )
+    }
+  }
+
+  /**
+   * Sends an API request over the web socket connection.
+   *
+   * @param request The object to send over the web socket.
+   * @returns The API response from the web socket.
+   */
+  private async sendApiRequest(
+    request: WebSocketRequestOptions,
+  ): Promise<WebSocketStatusResponse> {
+    while (this.socket.readyState === 0) {
+      await sleep(5)
+    }
+    if (this.socket.readyState !== 1) {
```

Comment:
can you make an enum for the states? does one exist already in the websocket module?

Reply:
Couldn't find one quickly so I just made one